pdr_backend/pdr_dashboard/dash_components/plots.py

```diff
@@ -345,38 +345,63 @@ def process_payouts(
     """
     processed = ProcessedPayouts()
 
-    p = payouts.copy()
+    p = payouts.clone()
     processed.predictions = len(payouts)
-    p["predictions_crt"] = range(1, len(payouts) + 1)
+
+    p = p.with_columns(
+        pl.lit(1).alias("predictions_crt"),
+        pl.struct(["payout", "stake"])
+        .map_elements(lambda x: max(x["payout"], 0) - x["stake"], pl.Float64)
+        .alias("profit_change"),
+    )
+    p = p.with_columns(
+        pl.col("predictions_crt").cum_sum().alias("predictions_crt"),
+        pl.when(pl.col("payout") > 0).then(1).otherwise(0).alias("correct_prediction"),
+        pl.when(pl.col("payout") > 0)
+        .then(1)
+        .otherwise(0)
+        .cum_sum()
+        .alias("correct_predictions_crt"),
+        pl.col("slot")
+        .map_elements(lambda x: UnixTimeS(int(x)).to_milliseconds(), pl.Int64)
+        .alias("slot_in_unixts"),
+    )
     processed.tx_cost = tx_fee_cost * len(payouts)
-    p["profit_change"] = p.apply(
-        lambda x: max(x["payout"], 0) - x["stake"], axis=1
-    ).astype(float)
-    p["correct_prediction"] = p["payout"].apply(lambda x: x > 0).astype(int)
-    p["correct_predictions_crt"] = p["correct_prediction"].cumsum()
+
     processed.correct_predictions = p["correct_prediction"].sum()
 
     if calculate_confint:
-        series = p.apply(
-            lambda x: proportion_confint(
-                count=x["correct_predictions_crt"], nobs=x["predictions_crt"]
-            ),
-            axis=1,
-        ).apply(pandas.Series)
-
-        processed.acc_l = series[0]
-        processed.acc_u = series[1]
-
-    processed.slot_in_unixts = p["slot"].apply(
-        lambda x: UnixTimeS(int(x)).to_milliseconds()
-    )
-    processed.accuracies = p.apply(
-        lambda x: x["correct_predictions_crt"] / x["predictions_crt"] * 100, axis=1
+        df = p.with_columns(
+            pl.struct(["correct_predictions_crt", "predictions_crt"])
+            .map_elements(
+                lambda x: proportion_confint(
+                    count=x["correct_predictions_crt"], nobs=x["predictions_crt"]
+                ),
+                return_dtype=pl.List(pl.Float64),
+            )
+            .alias("acc_bounds")
+        )
+
+        df = df.with_columns(pl.col("acc_bounds").list.get(0).alias("acc_l"))
+        df = df.with_columns(pl.col("acc_bounds").list.get(1).alias("acc_u"))
+
+        processed.acc_l = df["acc_l"]
+        processed.acc_u = df["acc_u"]
```

Comment:
We can avoid from using with_columns unnecessary

```
df = p.with_columns(
    [
        pl.struct(["correct_predictions_crt", "predictions_crt"])
        .map_elements(
            lambda x: proportion_confint(
                count=x["correct_predictions_crt"], nobs=x["predictions_crt"]
            ),
            return_dtype=pl.List(pl.Float64),
        )
        .alias("acc_bounds"),
        pl.col("acc_bounds").list.get(0).alias("acc_l"),
        pl.col("acc_bounds").list.get(1).alias("acc_u")
    ]
)

# Assigning the results to `processed` in one step
processed.acc_l = df["acc_l"]
processed.acc_u = df["acc_u"]
```

Reply:
doesn't work because "acc_bounds" is only defined as part of the first "with_columns". I can't split the list in the same one. I can only do it in a second "with_columns". Or, if you have any other ideas I'm happy to try something else

Reply:
you are right